R/logit.R

```diff
@@ -50,51 +50,44 @@ mle_BFGS_logit <- function(design, outcome) {
   )
 }
 
-#' Response probability
-#'
-#' Calculate the response probability with the design and coefficient in the logit model.
-#'
-#' @details
-#'
-#' @param design matrix, the design matrix X for the predictors.
-#' @param beta vector, the linear coefficients.
-#'
-#' @return A probability vector for the response.
-#'
 outcome_est <- function(design, beta) {
   1 - 1 / (1 + exp(crossprod(t(design), beta)))
 }
 
-#' Update step for beta
-#'
-#' Calculate the update step for the linear coefficient beta using Newton's method.
-#'
-#' @details
-#'
-#' @param grad vector, the gradient of the logit likelihood.
-#' @param Hessian matrix, the Hessian matrix of the likelihood
-#'
-#' @return A moving step for the linear coefficient beta.
-#'
 update_step <- function(grad, Hessian) {
   step <- solve(Hessian, grad)
   return(step)
 }
 
-#' Hessian matrix calculation
-#'
-#' Calculate the Hessian matrix for the current beta.
-#'
-#' @details
-#'
-#' @param design matrix, the design matrix X for the predictors.
-#' @param beta vector, the linear coefficients.
-#'
-#' @return The Hessian matrix.
-#'
-make_Hessian <- function(design, beta) {
+make_W <- function(design, beta){
   p <- as.vector(1 / (1 + exp(crossprod(t(design), beta))))
-  Hessian <- -crossprod(design, p * design)
+  W <- p*(1-p)
+  return(W)
+}
+
+make_Hessian <- function(design, beta) {
+  W = make_W(design, beta)
+  Hessian <- -crossprod(design, W * design)
+}
+
+take_one_newton_step <-  function(design, outcome, beta, method = 'QR'){
+  W = make_W(design, beta)
+  p = outcome_est(design, beta)
+  X_temp = sqrt(W) * design
+  Y_temp = (outcome - p)/sqrt(W)
+  if (method == 'QR'){
+    step = QR_solve_rcpp(X_temp, Y_temp)
+  }
+  if (method == 'chol'){
+    step = chol_solve(X_temp, Y_temp)
+  }
+  beta_new <- beta + step
+
+  log_likelihood_new <- log_likelihood_logit(design, outcome, beta_new)
```

Comment:
I'd move this logic outside. Recall the idea that functions should do 1 thing. In this case this function's "one thing" is computing the new beta, which doesn't require computing the loglik. This also simplifies the return type